File: src/main/java/cz/vhromada/catalog/facade/GenreFacade.java
package cz.vhromada.catalog.facade;

import cz.vhromada.catalog.entity.Genre;
import cz.vhromada.result.Result;

/**
 * An interface represents facade for genres.
 *
 * @author Vladimir Hromada
 */
public interface GenreFacade extends CatalogParentFacade<Genre> {

    /**
     * Adds genre. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID isn't null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * </ul>
     *
     * @param data genre
     * @return result with validation errors
     */
    @Override
    Result<Void> add(Genre data);

    /**
     * Updates genre.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID is null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param data genre
     * @return result with validation errors
     */
    @Override
    Result<Void> update(Genre data);

}

File: src/main/kotlin/cz/vhromada/catalog/facade/GenreFacade.kt
package cz.vhromada.catalog.facade

import cz.vhromada.catalog.entity.Genre
import cz.vhromada.common.facade.MovableParentFacade
import cz.vhromada.common.result.Result

/**
 * An interface represents facade for genres.
 *
 * @author Vladimir Hromada
 */
interface GenreFacade : MovableParentFacade<Genre> {

    /**
     * Adds genre. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *
     * @param data genre
     * @return result with validation errors
     */
    override fun add(data: Genre): Result<Unit>

    /**
     * Updates genre.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Genre doesn't exist in data storage
     *
     * @param data genre
     * @return result with validation errors
     */
    override fun update(data: Genre): Result<Unit>

}
